### DISCUSSION TITLE
`chainid` opcode returns an error

### DISCUSSION THREAD

**Discussion by frank-lim-partior** [ConsenSys/quorum#1494] [Q&A] [answered]
#### System information

Geth version: `Network from this [commit](https://github.com/ConsenSys/quorum-dev-quickstart/commit/fa2c4f2038ebdf54ef962e5e7d44bfd444ebb1d8)`
OS & Version: Windows/Linux/OSX

#### Expected behaviour
Read chain id when `chainid()` opcode is called.

#### Actual behaviour
Error received when attempting to call this in remix connected to local quorum RPC:
`call to ChainIdCaller.getChainId errored: invalid opcode: CHAINID`

#### Steps to reproduce the behaviour
`pragma solidity ^0.7.0;

contract ChainIdCaller {
  function getChainId() public view returns (uint256) {
    uint256 id;
    // solhint-disable-next-line no-inline-assembly
    assembly {
        id := chainid()
    }
    return id;
  }
}`

1. Use remix to deploy the above contract in a local GoQuorum started from quickstart.
2. Invoke `getChainId()` from remix.

#### Backtrace

````
[backtrace]
````

When submitting logs: please submit them as text and not screenshots.

**Comment by achraf17** [comment, marked as answer]
Hi @frank-part ,
There are two solutions to fix the issue:

1. when compiling your contract, use the `--evm-version istanbul` flag [docs](https://docs.soliditylang.org/en/v0.8.15/using-the-compiler.html#setting-the-evm-version-to-target). You have the same config in remix under advanced configuration.
2. upgrade the solidity version to `^0.8.0` and you can use `block.chainid` directly without assembly code.
```js
function getChainId() public view returns (uint256) {
   return block.chainid;
}
```

You also need to ensure that you are running on the istanbul fork and that you have specified `istanbulBlock` in your genesis.

**Reply by frank-lim-partior** [reply]
Hi @achraf17, thank you for your reply. I just tried the same test with the option in remix. It still fails with the same message.

**Reply by antonydenyer** [reply]
What have you got in your genesis config? Specifically what have you got for `istanbulBlock`?

**Reply by frank-lim-partior** [reply]
I was using the vanilla config generated from https://github.com/ConsenSys/quorum-dev-quickstart/commit/fa2c4f2038ebdf54ef962e5e7d44bfd444ebb1d8.

```
# grep inside generated quorum-test-network directory
grep -r istanbulBlock * 
# outputs
config/goquorum/data/raft-standard-genesis.json:    "istanbulBlock": 0,
config/goquorum/data/raft-enhanced-genesis.json:    "istanbulBlock": 0,
```

**Reply by antonydenyer** [reply]
I assume you're running a raft network? I've tried this locally with a qbft network using - the only thing I needed to do was increase the gas in the scripts.

**Reply by frank-lim-partior** [reply]
I have tried 2 things:

1. Running on raft as configured in the test project
2. Running on a ibft network

**Reply by frank-lim-partior** [reply]
Please help to look into a solution for this as this suggestion did not work.

**Reply by baptiste-b-pegasys** [reply]
Can you try with this smart contract please

//SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.6.3;

```
contract ChainId {

    constructor() public {
    }

    function get() public view returns (uint256) {
        uint256 id;
        assembly {
            id := chainid()
        }
        return id;
    }

}
```

**Reply by antonydenyer** [reply]
[chain-id-example.zip](https://github.com/ConsenSys/quorum/files/9607463/chain-id-example.zip)

Extract the example
`docker-compose up -d`
`cd smart_contracts`
`npm install`
`npm start`

let me know how get on

**Reply by frank-lim-partior** [reply]
@antonydenyer your example works, but uses `block.chainid`, which is a feature available in solidity 0.8.0.

This is different from the feature I'm describing, which is used in https://github.com/safe-global/safe-contracts/blob/main/contracts/GnosisSafe.sol#L340

**Reply by frank-lim-partior** [reply]
@baptiste-b-pegasys I tried updating the same project provided by antony with this code.
This works with the provided `docker-compose.yml` but it still fails with the same error on the quorum-test-network (and our environment).
[smart_contracts.zip](https://github.com/ConsenSys/quorum/files/9612434/smart_contracts.zip)
1. run `npx quorum-dev-quickstart` according to https://consensys.net/docs/goquorum/en/stable/tutorials/quorum-dev-quickstart/using-the-quickstart/
2. `cd quorum-test-network`
3. `./run.sh`
4. extract `smart_contracts.zip`
5. `cd smart_contracts`
6. `npm install`
7. `npm start`

Here are the `.env` and `docker-compose.yml` that was generated for my test network:

[config_in_quorum-test-network.zip](https://github.com/ConsenSys/quorum/files/9612490/config_in_quorum-test-network.zip)

**Reply by frank-lim-partior** [reply]
Here is the 'config' folder generated by `quorum-test-network`
[goquorum.zip](https://github.com/ConsenSys/quorum/files/9612542/goquorum.zip)

**Reply by antonydenyer** [reply]
You need to specify the block height at which to apply the istanbul fork which contains the opcode for requesting the chainid ie you need to configure`istanbulBlock` - I've updated the original answer to clarify for others.

**Reply by frank-lim-partior** [reply]
Hi @antonydenyer, can you provide an example or recommend a specific change in my config file? 
I don't see any specific mention of `istanbulBlock` in the example chaind-id-example you have given, where the example was working.

**Reply by antonydenyer** [reply]
In `config/goquorum/data/genesis.json` `istanbulBlock: 0` is where it's set. The block number should be higher than your current block if applying to an existing network. You need to run `geth init` after you make the change.  In your case you need to add this entry to your genesis files, assuming you are using the defaults you will be using qbft, you need to add `istanbulBlock` into the config property in both `qbft-standard-genesis.json` and `qbft-enhanced-genesis.json`

**Reply by frank-lim-partior** [reply]
Thank you - this is okay now.
